### DISCUSSION TITLE
Baking navigation meshes should not require MeshInstances to be children of NavigationMeshInstance

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#4879] [3D]
Godot has a powerful algorithm for baking NavigationMesh instances. To bake a navigation mesh the user needs to create a sub tree with a Navigation mesh at the root, a NavigationMeshInstance as its child and then MeshInstances as children of that.  The user then selects the NavigationMeshInstance  and clicks the Bake button that appears in the toolbar.  This will analyze the child MeshInstances and generate a navigation mesh resource based on that.

Unfortunately, there does not seem to be a way to bake imported scenes.  Even if I make my imported scene the child of the NavigationMeshInstance, the bake fails with an error.  I'm guessing this is because the root of any imported scene is a Spatial and the baking algorithm does not look recursively into the scene for buried MeshInstances.  

It would be helpful if the bake could handle scenes with deeply buried MeshInstances like this.  In fact, it would be helpful if the MeshInstances did not have to be descendants of the NavigationMeshInstance at all.  There is no reason for them to be children of the Navigation node other than that the bake script currently requires it.  The bake script should be changed to allow the user to specify any node anywhere in the scene to be the root of the bake.

**Comment by xix-xeaon** [comment]
In 3.5 you can select meshes/static colliders by group and they don't need to be children.

**Comment by Calinou** [comment]
cc @smix8

**Reply by smix8** [reply]
It is already possible to assign a groupname in the parse properties on the NavigationMesh resource before baking. All nodes inside this group will be parsed if they match the parse settings independent of their Scenetree position. There is nothing special about imported scenes that would prevent a navigationmesh parse as described apart from improper parse settings on the NavigationMesh resource for the geometry.

**Reply by blackears** [reply]
I'm using Godot 3.4.  Version 3.5 hasn't been released yet.  Also, how would I add the nodes of an imported scene to a group?

**Reply by smix8** [reply]
This is not a new feature from Godot 3.5 or 4.0. NavigationMesh bake groups have been part of Godot for a while but runtime (re)baking of navigationmesh only works in Godot 4.0/3.5. For Godot 3.4 your options are to use a custom import script to add groups in the asset import stage or use a tool script to add groups to scenes while in the Editor.

**Reply by blackears** [reply]
How do you tell the bake tool which group to use?

**Reply by smix8** [reply]
On the NavigationMeshInstance open the NavigationMesh resource in the inspector. You find the parse settings under the "Geometry" group. First is "Parsed Geometry Type" that sets if meshinstances or static colliders or both should be parsed. Second is the collision mask for static colliders (if used), third is the "Source Geometry Mode" that sets if all child nodes, only child nodes of group members or only group members should be parsed (maybe not in that order, I am writing from memory). Last is the "Source Group Name" that sets the group name.
